snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+from __future__ import annotations
+
+import json
+from abc import ABC, abstractmethod
+from typing import (
+    Generic,
+    List,
+    MutableMapping,
+    Optional,
+    TypeVar,
+    TYPE_CHECKING,
+    Union,
+)
+
+from snuba_sdk.entity import Entity
+from snuba_sdk.conditions import Condition
+from snuba_sdk.expressions import (
+    Column,
+    Function,
+    Granularity,
+    Limit,
+    LimitBy,
+    Offset,
+    OrderBy,
+    Totals,
+)
+from snuba_sdk.visitors import Translation
+
+if TYPE_CHECKING:
+    from snuba_sdk.query import Query
+
+
+class InvalidQuery(Exception):
+    pass
+
+
+QVisited = TypeVar("QVisited")
+
+
+class QueryVisitor(ABC, Generic[QVisited]):
+    def visit(self, query: Query) -> QVisited:
+        returns = {
+            "dataset": self._visit_dataset(query.dataset),
+            "match": self._visit_match(query.match),
+            "select": self._visit_select(query.select),
+            "groupby": self._visit_groupby(query.groupby),
+            "where": self._visit_where(query.where),
+            "having": self._visit_having(query.having),
+            "orderby": self._visit_orderby(query.orderby),
+            "limitby": self._visit_limitby(query.limitby),
+            "limit": self._visit_limit(query.limit),
+            "offset": self._visit_offset(query.offset),
+            "granularity": self._visit_granularity(query.granularity),
+            "totals": self._visit_totals(query.totals),
+        }
+
+        return self._combine(query, returns)
+
+    @abstractmethod
+    def _combine(
+        self,
+        query: Query,
+        returns: MutableMapping[str, QVisited],
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_dataset(self, dataset: str) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_match(self, match: Entity) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_select(
+        self, select: Optional[List[Union[Column, Function]]]
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_groupby(
+        self, groupby: Optional[List[Union[Column, Function]]]
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_where(self, where: Optional[List[Condition]]) -> QVisited:
```

Comment:
Why a list and not a Sequence? (list is mutable, sequence is not).

Reply:
These match the types in the Query object, which specifies a list. Passing a tuple to this function would be incorrect.

Reply:
ok